# Arcadia-Science/prehgt: Citations

A number of tools were used to make the prehgt pipeline.
Depending on how you use the prehgt pipeline, please consider citing the papers and resources below.

## Workflow tools

- [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)
  > Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. (2020). The nf-core framework for community-curated bioinformatics pipelines. https://doi.org/10.1038/s41587-020-0439-x

- [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)
  > Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. (2017). Nextflow enables reproducible computational workflows. https://doi.org/10.1038/nbt.3820

- [Snakemake](https://snakemake.readthedocs.io/en/stable/)
  > Köster J, Rahmann S. (2012). Snakemake—a scalable bioinformatics workflow engine. https://doi.org/10.1093/bioinformatics/bts480

## Pipeline tools

- [csvtk](https://github.com/shenwei356/csvtk)

- [DIAMOND](https://github.com/bbuchfink/diamond)
  > Buchfink B, Xie C, Huson DH. (2014). Fast and sensitive protein alignment using DIAMOND. https://doi.org/10.1038/nmeth.3176

- [EMBOSS](https://emboss.sourceforge.net/)
  > Rice P, Longden I, Bleasby A. (2000). EMBOSS: the European molecular biology open software suite.

- [fastcluster](https://cran.r-project.org/web/packages/fastcluster/index.html)
  > Müllner D. (2013). fastcluster: Fast Hierarchical, Agglomerative Clustering. https://doi.org/10.18637/jss.v053.i09

- [HMMER3](http://hmmer.org/)

- [KofamScan](https://github.com/takaram/kofam_scan)
  > Aramaki T, Blanc-Mathieu R, Endo H, Ohkubo K, Kanehisa M, Goto S, Ogata H. (2020). KofamKOALA: KEGG Ortholog assignment based on profile HMM and adaptive score threshold. https://doi.org/10.1093/bioinformatics/btz859

- [MMSeq2](https://github.com/soedinglab/MMseqs2)
  > Steinegger M, Söding J. (2017). MMseqs2 enables sensitive protein sequence searching for the analysis of massive data sets. https://doi.org/10.1038/nbt.3988

- [ncbi-genome-download](https://github.com/kblin/ncbi-genome-download)

- [seqtk](https://github.com/lh3/seqtk)

- [tidyverse](https://www.tidyverse.org/)
  > Wickham H, Averick M, Bryan J, Chang W, McGowan LD, François R, Grolemund G, Hayes A, Henry L, Hester J, Kuhn M. (2019). Welcome to the Tidyverse. https://doi.org/10.21105/joss.01686

## Horizontal gene transfer prediction algorithms

- Alien index
  > Gladyshev EA, Meselson M, Arkhipova IR. (2008). Massive horizontal gene transfer in bdelloid rotifers. https://doi.org/10.1126/science.1156407

- Donor distribution index
  > Ma J, Wang S, Zhu X, Sun G, Chang G, Li L, Hu X, Zhang S, Zhou Y, Song CP, Huang J. (2022). Major episodes of horizontal gene transfer drove the evolution of land plants. https://doi.org/10.1016/j.molp.2022.02.001

- HGT score
  > Boschetti C, Carr A, Crisp A, Eyres I, Wang-Koh Y, Lubzens E, Barraclough TG, Micklem G, Tunnacliffe A. (2012). Biochemical diversification through foreign gene expression in bdelloid rotifers. https://doi.org/10.1371/journal.pgen.1003035

- Transfer index
  > Nguyen M, Ekstrom A, Li X, Yin Y. (2015). HGT-Finder: A new tool for horizontal gene transfer finding and application to Aspergillus genomes. https://doi.org/10.3390/toxins7104035

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)
  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)
  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. (2018). Bioconda: sustainable and comprehensive software distribution for the life sciences. https://doi.org/10.1038/s41592-018-0046-7

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)
  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. (2017). BioContainers: an open-source and community-driven framework for software standardization. https://doi.org/10.1093/bioinformatics/btx192
